Mark software API endpoints as experimental

marko-lisica opened the pull request:
- Mark new API endpoints or API endpoints that were changed as part of Fleet's first app management feature (#14921) as beta. 
- Call out what is in beta exactly (the endpoint or new keys/values) and point to changes

marko-lisica commented:
Hey @lukeheath and @noahtalerman. Here's the PR to mark software endpoints beta.

I left out [List software versions](https://fleetdm.com/docs/rest-api/rest-api#list-software-versions) and [Get software version](https://fleetdm.com/docs/rest-api/rest-api#get-software-version). There are existed for a long time and they are not affected by the new features we're adding.

noahtalerman commented:
Thanks @marko-lisica! Realizing now that many customers (read-only) integrate w/ the software API endpoints but aren't using the new app management features. 

So, I think it's important to be specific that the app management parts of the API are subject to change and link to what those changes are. That way, the read-only customers can look at the API and say "ok, this won't affect me."

Working on making these changes.

cc @lukeheath

noahtalerman commented:
Hey @marko-lisica, I updated the PR to only call out the endpoints that are breaking.

Please let me know if I missed an endpoint!

I'm also calling out if the endpoint URL is breaking or if values/keys are breaking.

I added links to the upcoming changes.

lukeheath commented on docs/REST API/rest-api.md:
```diff
@@ -4276,6 +4276,8 @@ OS vulnerability data is currently available for Windows and macOS. For other pl
 
 ### Get host's software
 
+> The [new keys/values](https://github.com/fleetdm/fleet/pull/19245/files) added in the app management features are in beta and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291).
```
I think we should be more explicit than linking to changes. What is there now, and what is it changing to?

lukeheath replied:
Or we should at least link to the specific lines we are referencing. Otherwise very hard to understand what is being references at those links.

noahtalerman replied:
@lukeheath great points. My latest commit pulls out links to the old (merged) PR and updates links to point to specific lines we are referencing.

marko-lisica commented on docs/REST API/rest-api.md:
```diff
@@ -4276,6 +4276,8 @@ OS vulnerability data is currently available for Windows and macOS. For other pl
 
 ### Get host's software
 
+> The new keys/values added in the app management features are in beta and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L4304-L4311).
```
```suggestion
> The new keys/values added in the software management features are in beta and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L4304-L4311).
```

marko-lisica replied:
@noahtalerman I think we don't use "app management" anywhere else. We always use "software" except for App Store apps.

noahtalerman replied:
Hey @marko-lisica good catch.

We're starting to use "App management" on the website.

Check out [fleetdm.com/device-management](https://fleetdm.com/device-management):
![Screenshot 2024-07-24 at 3 02 06 PM](https://github.com/user-attachments/assets/04d95ac4-d274-4f01-aa02-a5bd87870238)

marko-lisica commented:
> Hey @marko-lisica, I updated the PR to only call out the endpoints that are breaking.
> 
> Please let me know if I missed an endpoint!
> 
> I'm also calling out if the endpoint URL is breaking or if values/keys are breaking.
> 
> I added links to the upcoming changes.

@noahtalerman looks good to me. I just left some small copy suggestions.

marko-lisica commented on docs/REST API/rest-api.md:
```diff
@@ -8656,6 +8658,8 @@ Body: <blob>
 
 ### Delete software
 
+> This endpoint, added in the app management feature, is in beta and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L8661-R8698).
```
```suggestion
> This endpoint, added in the software management feature, is in beta and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L8661-R8698).
```

marko-lisica commented on docs/REST API/rest-api.md:
```diff
@@ -8714,6 +8718,8 @@ Get the results of a software installation.
 
 ### List software
 
+> The new keys/values added in the app management feature are in beta and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L8749-R8791).
```
```suggestion
> The new keys/values added in the software management feature are in beta and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L8749-R8791).
```

lukeheath commented on docs/REST API/rest-api.md:
```diff
@@ -4276,6 +4276,8 @@ OS vulnerability data is currently available for Windows and macOS. For other pl
 
 ### Get host's software
 
+> The new keys/values added in the app management features are experimental and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L4304-L4311).
```
```suggestion
> **This feature is still experimental.** 
>
>The new keys/values added in the app management features are experimental and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L4304-L4311).
```

What do you think of making this callout clearer like Mike's example from Sails?

lukeheath commented on docs/REST API/rest-api.md:
```diff
@@ -8656,6 +8658,8 @@ Body: <blob>
 
 ### Delete software
 
+> This endpoint, added in the app management feature, is experimental and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L8661-R8698).
```
```suggestion
> **This feature is still experimental.** 
>
> This endpoint, added in the app management feature, is experimental and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L8661-R8698).
```

lukeheath commented on docs/REST API/rest-api.md:
```diff
@@ -8714,6 +8718,8 @@ Get the results of a software installation.
 
 ### List software
 
+> The new keys/values added in the app management feature are experimental and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L8749-R8791).
```
```suggestion
> **This feature is still experimental.** 
>
> The new keys/values added in the app management feature are experimental and may change. You can find the upcoming breaking changes [here](https://github.com/fleetdm/fleet/pull/19291/files#diff-7246bc304b15c8865ed8eaa205e9c244d0a0314e4bae60cf553dc06147c38b64L8749-R8791).
```

noahtalerman replied:
I think I see what you're after but I think the purple call out box is loud enough?

I added bold to "experimental"

Not sure how the purple callout box would render with a bold title/header...

lukeheath reviewed (approved):
Bold works 👍
